chalice/package.py

```diff
@@ -594,3 +605,67 @@ def _fixup_deployment_package(self, template, outdir):
                 self._osutils.copy(original_location, new_location)
                 copied = True
             resource['Properties']['CodeUri'] = './deployment.zip'
+
+
+class TemplateMergePostProcessor(TemplatePostProcessor):
+    def process(self, template, config, outdir, chalice_stage_name):
+        # type: (Dict[str, Any], Config, str, str) -> None
+        if not self._should_merge_template(config):
+            return
+        loaded_template = self._load_template_to_merge(config)
+        merger = TemplateDeepMerger()
+        merged = merger.merge(loaded_template, template)
+        template.clear()
+        template.update(merged)
+
+    def _should_merge_template(self, config):
+        # type: (Config) -> bool
+        return config.package_merge_template is not None
+
+    def _load_template_to_merge(self, config):
+        # type: (Config) -> Dict[str, Any]
+        filepath = os.path.abspath(config.package_merge_template)
+        if not self._osutils.file_exists(filepath):
+            raise RuntimeError('Cannot find template file: %s' % filepath)
+        template_data = self._osutils.get_file_contents(filepath, binary=False)
+        try:
+            loaded_template = json.loads(template_data)
+        except ValueError:
+            raise RuntimeError(
+                'Expected %s to be valid JSON template.' % filepath)
+        return loaded_template
+
+
+class TemplateDeepMerger(object):
+    def merge(self, src, dst):
+        # type: (Dict[str, Any], Dict[str, Any]) -> Dict[str, Any]
+        return self._merge_dict(src, dst)
+
+    def _merge(self, src, dst):
+        # type: (Dict[str, Any], Dict[str, Any]) -> Dict[str, Any]
+        src_type = type(src).__name__
+        dst_type = type(dst).__name__
+        if src_type != dst_type:
+            name = '_merge_unknown'
+        else:
+            name = '_merge_%s' % src_type
+        return getattr(self, name, self._merge_replace)(src, dst)
+
+    def _merge_dict(self, src, dst):
+        # type: (Dict[str, Any], Dict[str, Any]) -> Dict[str, Any]
+        result = dst.copy()
+        for key, value in src.items():
+            existing_value = dst.get(key)
+            if existing_value is None:
+                result[key] = value
+            else:
+                result[key] = self._merge(value, existing_value)
+            if result[key] is None:
```

Comment:
Why do we want this?  I think this no longer makes this a "deep merge" if certain values have special semantics.

Reply:
Yea I was immediately thinking of the case where someone wants to remove a key added by Chalice not replace it. I do agree its odd for the default behavior, I also can't think of a great concrete reason for it, it just seemed like an odd gap.